File: src/SnakesandLadders.java
import java.util.LinkedList;
import java.util.Queue;

/*
 *  The Time complexity of the algorithm is O(m*n)) where m*n is no of elements in matrix.
 *  Space complexity of the algorithm is O(n) .
 */

public class SnakesandLadders {

	public int snakesAndLadders(int[][] board) {

		int r = board.length;
		int[] moves = new int[r * r];

		int idx = 0, j = 0, i = r - 1;
		int even = 0;

		while (i >= 0 && j >= 0) {

			if (board[i][j] == -1) {

				moves[idx] = -1;
			} else {
				moves[idx] = board[i][j] - 1;
			}

			idx++;

			if (even % 2 == 0) {
				j++;

			} else {

				j--;
			}

			if (j == r) {
				even++;
				j--;
				i--;

			}

			if (j < 0) {
				even++;
				j++;
				i--;

			}

		}

		Queue<Integer> q = new LinkedList<>();
		int min = 0;
		q.add(0);
		moves[0] = -2;

		while (!q.isEmpty()) {

			int size = q.size();

			for (int k = 0; k < size; k++) {

				int curr = q.poll();

				if (curr == moves.length - 1)
					return min;

				for (int l = 1; l <= 6; l++) {

					int child = curr + l;

					if (child <= moves.length - 1 && moves[child] != -2) {
						if (moves[child] == -1) {

							q.add(child);
						} else {
							q.add(moves[child]);
						}

						moves[child] = -2;
					}

				}

			}

			min++;

		}

		return -1;
	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}

File: Snakes And Ladders.py
# Time complexity : O(n^2)
# Space complexity : O(n^2)
class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        if not board or not board[0]:
            return 0
        ans = 0
        visited = set()
        n = len(board)
        flattened = []
        flag = False
        
        idx = 0
        for i in range(n-1,-1,-1):
            if not flag:
                for j in range(n):
                    if board[i][j] >-1:
                        flattened.append(board[i][j]-1)
                    else:
                        flattened.append(idx)
                    idx += 1
            else:
                for j in range(n-1,-1,-1):
                    if board[i][j] >-1:
                        flattened.append(board[i][j]-1)
                    else:
                        flattened.append(idx)
                    idx += 1
            if flag:
                flag = False
            else:
                flag = True
        
        q = collections.deque()
        
        q.append(flattened[0])
        visited.add(flattened[0])
        
        moves = 0
        
        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                if curr == (n*n -1):
                    return moves
                limit = min(curr+6,n*n-1)
                for j in range(curr+1,limit+1):
                    if flattened[j] not in visited:
                        if flattened[j] == (n*n -1):
                            return moves+1
                        q.append(flattened[j])
                        visited.add(flattened[j])
            moves += 1
        
        return -1
                    
                    
                    
                